improver_tests/test_field_texture.py

```diff
@@ -0,0 +1,244 @@
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+""" Tests of FieldTexture plugin."""
+
+import iris
+import numpy as np
+import pytest
+from iris.exceptions import CoordinateNotFoundError
+
+from improver.field_texture import FieldTexture
+from improver.synthetic_data.set_up_test_cubes import (
+    add_coordinate,
+    set_up_probability_cube,
+)
+
+NB_RADIUS = 10000
+TEXT_THRESH = 0.4
+DIAG_THRESH = 0.8125
+PLUGIN = FieldTexture(
+    nbhood_radius=NB_RADIUS,
+    textural_threshold=TEXT_THRESH,
+    diagnostic_threshold=DIAG_THRESH,
+)
+
+
+@pytest.fixture(name="multi_cloud_cube")
+def multi_cloud_fixture():
+    """Multi-realization/threshold cloud data cube."""
+    cloud_area_fraction = np.zeros((3, 10, 10), dtype=np.float32)
+    cloud_area_fraction[:, 1:4, 1:4] = 1.0
+    thresholds = [0.265, 0.415, 0.8125]
+    return cloud_probability_cube(cloud_area_fraction, thresholds)
+
+
+@pytest.fixture(name="thresholded_cloud_cube")
+def thresholded_cloud_fixture():
+    """Cloud data for a single realization at the relevant threshold."""
+    cloud_area_fraction = np.zeros((1, 10, 10), dtype=np.float32)
+    cloud_area_fraction[:, 1:4, 1:4] = 1.0
+    thresholds = [DIAG_THRESH]
+    multi_realization_cube = iris.util.squeeze(
+        cloud_probability_cube(cloud_area_fraction, thresholds)
+    )
+    return next(multi_realization_cube.slices_over("realization"))
+
+
+@pytest.fixture(name="no_cloud_cube")
+def no_cloud_fixture():
+    """Multi-realization cloud data cube with no cloud present."""
+    cloud_area_fraction = np.zeros((3, 10, 10), dtype=np.float32)
+    thresholds = [0.265, 0.415, 0.8125]
+    return cloud_probability_cube(cloud_area_fraction, thresholds)
+
+
+@pytest.fixture(name="all_cloud_cube")
+def all_cloud_fixture():
+    """Multi-realization cloud data cube with all cloud present."""
+    cloud_area_fraction = np.ones((3, 10, 10), dtype=np.float32)
+    thresholds = [0.265, 0.415, 0.8125]
+    return cloud_probability_cube(cloud_area_fraction, thresholds)
+
+
+# Set up probability cubes for the pytest fixtures.
+
+
+def cloud_probability_cube(cloud_area_fraction, thresholds):
+    """Set up probability cube."""
+    cube = set_up_probability_cube(
+        cloud_area_fraction,
+        variable_name="cloud_area_fraction",
+        threshold_units="1",
+        thresholds=thresholds,
+        spatial_grid="equalarea",
+    )
```

Comment:
There is just one more thing we need to test for (sorry!), and that is attributes.  To do that, you'll need to make some changes to the input cubes here, and then check the result later.

Could you please use the `attributes` argument to `set_up_probability_cube` with value:
```
attributes = {
    'source': 'Unified Model',
    'institution': 'Met Office',
    'title': 'MOGREPS-UK Forecast on 2 km Standard Grid',
    'mosg__model_configuration': 'uk_ens' 
}
```
An additional metadata unit test should then check for the following expected attributes on the output cube:
```
expected_attributes = {
    'source': 'Unified Model',
    'institution': 'Met Office',
    'title': 'MOGREPS-UK Forecast on 2 km Standard Grid'
}
```
(This isn't a typo - only these three attributes should be inherited.)

Reply:
Done.